src/native_mux_impl/stdio.rs

```diff
@@ -28,25 +29,50 @@ fn get_null_fd() -> Result<RawFd, Error> {
 }
 
 pub(crate) enum Fd {
-    PipeReadEnd(PipeRead),
-    PipeWriteEnd(PipeWrite),
-
+    Owned(OwnedFd),
     Borrowed(RawFd),
     Null,
 }
 
+fn cvt(ret: c_int) -> io::Result<c_int> {
+    if ret == -1 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(ret)
+    }
+}
+
+fn set_blocking(fd: RawFd) -> io::Result<()> {
+    let flags = cvt(unsafe { fcntl(fd, F_GETFL) })?;
+    cvt(unsafe { fcntl(fd, F_SETFL, flags & (!O_NONBLOCK)) })?;
+
+    Ok(())
+}
+
 impl Fd {
     pub(crate) fn as_raw_fd_or_null_fd(&self) -> Result<RawFd, Error> {
         use Fd::*;
 
-        match self {
-            PipeReadEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
-            PipeWriteEnd(fd) => Ok(AsRawFd::as_raw_fd(fd)),
+        let fd = match self {
+            Owned(owned_fd) => Some(owned_fd.as_raw_fd()),
+            Borrowed(rawfd) => Some(*rawfd),
+            Null => None,
+        };
 
-            Borrowed(rawfd) => Ok(*rawfd),
-            Null => get_null_fd(),
+        if let Some(fd) = fd {
+            set_blocking(fd).map_err(Error::ChildIo)?;
+
+            Ok(fd)
+        } else {
+            get_null_fd()
         }
     }
+
+    fn new_owned<T: IntoRawFd>(fd: T) -> Self {
+        let raw_fd = fd.into_raw_fd();
+        // Safety: IntoRawFd::into_raw_fd must return a valid raw fd.
```

Comment:
That's not true. [The docs](https://doc.rust-lang.org/std/os/unix/io/trait.IntoRawFd.html#tymethod.into_raw_fd) explicitly say it's not a requirement of `IntoRawFd` and recommend another way to do it.

Reply:
Unfortunately it doesn't work since `tokio_pipe::{PipeRead, PipeWrite}` does not implement `Into<OwnedFd>`

Reply:
In that case we should not take a generic `T` here, but instead have constructors _specifically_ for `PIpeRead` and `PipeWrite` where we know that _their_ implementations of `IntoRawFd` _do_ guarantee that ownership is transferred.

Reply:
Done!

To avoid duplicate code, I make `Fd::new_owned` an unsafe function and then implements `impl From<PipeRead> for Fd` and `impl From<PipeWrite> for Fd`.